File: Go/ConcatenatedWords.go
//go:build ignore

package main

import "fmt"

func findAllConcatenatedWordsInADict(words []string) []string {
	set := make(map[string]struct{}, len(words))
	for _, w := range words {
		set[w] = struct{}{}
	}

	canForm := func(word string) bool {
		n := len(word)
		dp := make([]bool, n+1)
		dp[0] = true
		for i := 1; i <= n; i++ {
			for j := 0; j < i; j++ {
				if !dp[j] || (j == 0 && i == n) {
					continue
				}
				if _, ok := set[word[j:i]]; ok {
					dp[i] = true
					break
				}
			}
		}
		return dp[n]
	}

	res := []string{}
	for _, w := range words {
		if w != "" && canForm(w) {
			res = append(res, w)
		}
	}
	return res
}

func main() {
	fmt.Println(findAllConcatenatedWordsInADict([]string{"cat", "cats", "catsdogcats", "dog",
		"dogcatsdog", "hippopotamuses", "rat", "ratcatdogcat"}))
	// [catsdogcats dogcatsdog ratcatdogcat]
	fmt.Println(findAllConcatenatedWordsInADict([]string{"cat", "dog", "catdog"})) // [catdog]
	fmt.Println(findAllConcatenatedWordsInADict([]string{"a", "b", "ab", "abc"}))  // [ab]
}

File: Java/ConcatenatedWords.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class ConcatenatedWords {
    public List<String> findAllConcatenatedWordsInADict(String[] words) {
        Set<String> set = new HashSet<>();
        for (String w : words) set.add(w);
        List<String> res = new ArrayList<>();
        for (String w : words) {
            if (!w.isEmpty() && canForm(w, set)) res.add(w);
        }
        return res;
    }

    private boolean canForm(String word, Set<String> set) {
        int n = word.length();
        boolean[] dp = new boolean[n + 1];
        dp[0] = true;
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < i; j++) {
                if (!dp[j] || (j == 0 && i == n)) continue;
                if (set.contains(word.substring(j, i))) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[n];
    }

    public static void main(String[] args) {
        ConcatenatedWords s = new ConcatenatedWords();
        System.out.println(s.findAllConcatenatedWordsInADict(new String[] { "cat", "cats", "catsdogcats",
                "dog", "dogcatsdog", "hippopotamuses", "rat", "ratcatdogcat" }));
        System.out.println(s.findAllConcatenatedWordsInADict(new String[] { "cat", "dog", "catdog" }));
        System.out.println(s.findAllConcatenatedWordsInADict(new String[] { "a", "b", "ab", "abc" }));
    }
}
